ml-agents/mlagents/trainers/sac/models.py

```diff
@@ -0,0 +1,916 @@
+import logging
+import numpy as np
+
+import tensorflow as tf
+from mlagents.trainers.models import LearningModel
+import tensorflow.contrib.layers as c_layers
+
+LOG_STD_MAX = 2
+LOG_STD_MIN = -20
+EPSILON = 1e-6  # Small value to avoid divide by zero
+
+logger = logging.getLogger("mlagents.trainers")
+
+POLICY_SCOPE = ""
+TARGET_SCOPE = "target_network"
+
+
+class SACNetwork(LearningModel):
+    def __init__(
+        self,
+        brain,
+        m_size=None,
+        h_size=128,
+        normalize=False,
+        use_recurrent=False,
+        num_layers=2,
+        stream_names=None,
+        seed=0,
+        is_target=False,
+        vis_encode_type="default",
+    ):
+        LearningModel.__init__(
+            self, m_size, normalize, use_recurrent, brain, seed, stream_names
+        )
+        self.normalize = normalize
+        self.use_recurrent = use_recurrent
+        self.num_layers = num_layers
+        self.stream_names = stream_names
+        self.h_size = h_size
+        self.share_ac_cnn = False
+        self.activ_fn = self.swish
+
+        if self.use_recurrent:
+            self.memory_in = tf.placeholder(
+                shape=[None, self.m_size], dtype=tf.float32, name="recurrent_in"
+            )
+
+        if is_target:
+            if self.use_recurrent:
+                self.value_memory_in = self.memory_in
+            with tf.variable_scope(TARGET_SCOPE):
+                hidden_streams = self.create_observation_streams(
+                    1,
+                    self.h_size,
+                    0,
+                    vis_encode_type=vis_encode_type,
+                    stream_scopes=["critic/value/"],
+                )
+            if brain.vector_action_space_type == "continuous":
+                self.create_cc_critic(hidden_streams[0], TARGET_SCOPE, create_qs=False)
+            else:
+                self.create_dc_critic(hidden_streams[0], TARGET_SCOPE, create_qs=False)
+            if self.use_recurrent:
+                mem_outs = [self.value_memory_out]
+        else:
+            if self.use_recurrent:
+                num_mems = 4
+                mem_ins = []
+                for i in range(num_mems):
+                    _start = self.m_size // num_mems * i
+                    _end = self.m_size // num_mems * (i + 1)
+                    mem_ins.append(self.memory_in[:, _start:_end])
+                self.value_memory_in = mem_ins[0]
+                self.q1_memory_in = mem_ins[1]
```

Comment:
Maybe a dumb question, but will we only ever want 2 networks in SAC?

Reply:
Even if not, it looks like some of the code could be cleaned up a bit by looping, for example create_losses

Reply:
Yes only ever 2 networks. The two networks are actually quite different in structure, the policy network has 4 sub networks (policy, value, q1, q2) vs. the target network which only has value. I think I can loop the creation on the upper level SAC model, though.